aries_cloudagent/revocation/routes.py

```diff
@@ -149,13 +150,51 @@ def validate_fields(self, data, **kwargs):
 class RevokeRequestSchema(CredRevRecordQueryStringSchema):
     """Parameters and validators for revocation request."""
 
+    @validates_schema
+    def validate_fields(self, data, **kwargs):
+        """Validate fields - connection_id and thread_id must be present if notify."""
+        super().validate_fields(data, **kwargs)
+
+        notify = data.get("notify")
+        connection_id = data.get("connection_id")
+        thread_id = data.get("thread_id")
+
+        if notify and not (connection_id and thread_id):
+            raise ValidationError(
+                "Request must specify thread_id and connection_id if notify is true"
+            )
+
     publish = fields.Boolean(
         description=(
             "(True) publish revocation to ledger immediately, or "
             "(default, False) mark it pending"
         ),
         required=False,
     )
+    notify = fields.Boolean(
+        description="Send a notification to the credential recipient",
+        required=False,
+    )
+    connection_id = fields.Str(
+        description=(
+            "Connection ID to which the revocation notification will be sent; "
+            "required if notify is true"
+        ),
+        required=False,
+        **UUID4,
+    )
+    thread_id = fields.Str(
+        description=(
+            "Thread ID of the credential exchange message thread resulting in "
+            "the credential now being revoked; required if notify is true"
+        ),
+        required=False,
+        **UUID4,
+    )
```

Comment:
I suggest that we make the `thread_id` optional when we're revoking (with notifications).  The cred_exchange records are transitory and the controller *may* not record the thread_id.  Likewise the holder may not know the `thread_id`.

Reply:
The RFC requires the thread ID. So, while I agree with your assessment that it's likely the thread_id is not known by one side or the other or both, this would be a spec divergent change. Happy to code it up and make it happen but wanted to call that out at least first.

Reply:
I think we probably should have reviewed this more carefully when we accepted the RFC.  :-(.

Obviously, for an Indy Credential, we would use a different ID to communicate what it is that we are revoking. For Indy we need a combination of RevRegID+IndexOfCredInRevReg. For any other credential format, there is likely some sort of ID available. I think the use of thread_id was picked because without knowing the details of a specific implementation, you can't guess what the ID should be.  Further, with the suggestion that we can issue multiple credentials using a single protocol instance (e.g. a bank sends to the holder a credential for each bank account they have, or a university sends a credential to the holder each degree they have been granted), the "thread_id" is not going to work.

I suggest we propose a change to the Aries RFC(s).  Ideal would be that the "Issue Credential" protocol define a "revocation_id" that is VC format specific and understood by the holder and issuer, and the Revocation Notification V2.0 message use that ID to the in the notification message.  Since I'd rather not change the "Issue Credential" protocol, I suggest that we change "thread_id" to "revocation_id", and define it to be VC format specific, and for indy have a concatenated key for the credential.  We could even say as a 1.0 clarification that the "thread_id" can be EITHER the thread id or the VC-format specific "revocation_id" that is defined in the V2.0 message.

Thoughts?

Reply:
We could define the format of what is to be included in `thread_id`, for example:

`"thread_id": "12345678-1234-1234-1234-123456789012"` = an actual thread id
`"thread_id": "indy::<reg_rev_id>::<cred_index_in_rev_reg>"` = an indy-specific format

This would at least respect the format of the notification message, if not (necessarily) the content.

However I agree with @swcurran that it's better to add a new `"revocation_id"` (could use the above format for indy) and push a clarification to the RFC (if "revocation_id" is specified then "thread_id" is not required)

Reply:
Agree with the addition of an opaque `revocation_id` used for this purpose. I have less of an opinion about how we bridge between old and new.

Reply:
I personally like Ian's suggestion, with the holder checking first to see if they have thread_id that matches what is passed, and then falling back to an implementation specific identifier.  AFAIK, this is a first implementation so I don't think it would be a huge issue.

 I vote we add a clarification to the RFP, and document this in v1.0, and also add either a v1.1 (keep thread_id and add revocation_id and how to process both) or v2.0 version (replace thread_id with revocation_id).
 
@smithbk -- thoughts on this?

Reply:
Here's my proposal for this PR:

- Make `thread_id` optional in ACA-Py's revoke endpoint. When absent and `notify` is `true`, ACA-Py will fill in and send a `thread_id` value in the revocation notification with the format `indy::<rev_reg_id>::<cred_index>`. When present, ACA-Py will use the given value.
- Loosen restrictions on format of `thread_id` to accommodate the above.
- Address `revocation_id` vs `thread_id` in a future revision once conversation has coalesced.

@ianco @swcurran does this seem like a good compromise as we drive towards better language in the RFC?

Reply:
Yes that makes sense to me